src/core/thread/mle.cpp

```diff
@@ -346,6 +347,8 @@ void Mle::SetRole(otDeviceRole aRole)
         mParent.SetState(Neighbor::kStateInvalid);
     }
 
+    OtnsStatus("role=%d", mRole);
```

Comment:
I am not in favor of emitting `OtnsStatus()` from within the OT core modules (as an example here or from `SubMac` for the ext address change, etc).  I think this can create dependencies to the tests that would be hard to maintain in future. The core code can/will change but we try to keep the external/public APIs as before and backwards-compatible. I think instead it would be better to use the public APIs and/or existing eventing mechanism to get info.

For example here, can we consider using the existing notifier `OT_CHANGED_THREAD_ROLE` event model?
An OTNS module can register a callback to `Notifier` (e.g., using `otSetStateChangedCallback()` or a direct register defining `Notifier::Callback`) and process the `OT_CHANGED_THREAD_ROLE` event  and emit the `"role=%d", mRole` from callback. With this approach the otns event may not be as immediate as a direct emit from `Mle::SetRole()`  (due to `Notifier` emitting its events from a posted `Tasklet`), but I don't expect it would matter for building test cases.

I would say if there are some internal states that we currently don't expose out, it would be better to add the appropriate APIs or events for those instead.

Thoughts?

Reply:
> 2.Notifier is using Tasklets which are posted rather than executed immediately. The virtual time simulator implemented in #2508 schedules all node events and transmissions to preserve the execution order in time. However, using Tasklets to send events breaks the strict ordering of events, thus could have impact on the correctness of simulation.

Can you please help explain why we need immediate eventing for OTNS to work (e.g., need to know exactly when RLOC16 or role value changes).
- The virtual time simulator in OT is emulating the platform layer and events are related to radio/timer, etc. 
- OpenThread core itself can be considered a big state machine which requires calls from platform layer and its public APIs to proceed, take any action and  move from one state to another.  
- So a simulator pretty much knows and is in control of when OT state changes (since it would be calling the platform/public APIs that cause OT to change). 
- A higher level simulator  can then use existing public APIs to query info/state from OT (e.g., check the current role using `otThreadGetDeviceRole()`). This can be done after every call that would change OT. 
- With such a model we should not need any immediate evening from OT. Can we  consider adopting such a model?

> The calls to OtnsStatus are placed exactly where callbacks are or should be. So, I would expect maintaining OtnsStatus calls simple because we will also have to maintain callback code as core code changes. We just need to make OtnsStatus follow corresponding callback signaling.

My reservation is for this exact purpose. For now the `OtnsStatus` may be placed where the callbacks are, but if we allow and use `OtnsStatus` directly in the core modules, it may be used in places where there is no callback and adds dependency in the code. 

I would highly suggest we funnel OTNS events through existing public APIs and/or adding new APIs/mechanism.

Reply:
> Can you please help explain why we need immediate eventing for OTNS to work (e.g., need to know exactly when RLOC16 or role value changes).

One of the motivation for OTNS to use status emitting to acquire node status rather than calling APIs is because the **node status affects the simulation result immediately**. 
For example, OTNS uses node `rloc16` and `extaddr` for destination address filtering so that it only dispatches unicast messages to the right destinations. This is one of the reasons that `OTNS` outperforms existing virtual time simulator by tens of times. 

Using `Tasklets` for event emitting changes the order of events, e.x. message transmission, extaddr/rloc16 change, etc.  If we are to implement MAC filters in OTNS, the order of message transmission events and `extaddr` change events could cause `OTNS` to incorrectly drop messages. 

In general, I am worried about the disordering of events caused by `posted Tasklets` because order of events impacts `OTNS` simulation result. 

> A higher level simulator can then use existing public APIs to query info/state from OT (e.g., check the current role using otThreadGetDeviceRole()). This can be done after every call that would change OT.

This however implies that status emitting to `OTNS` must be implemented in one of apps or platforms. However, I would expect `OTNS` to work for cross apps/platforms in the future. For example, with `OTNS` enhancements and proper `otPlatOtnsStatus` implementation on `nrf528xx` platform, we can even use `OTNS` for simulating or visualizing real devices. Integrating `OTNS` mechanisms in the OpenThread core equips `OTNS` with more potentials. 

> but if we allow and use OtnsStatus directly in the core modules, it may be used in places where there is no callback and adds dependency in the code.

I would expect this never to happen that we call `OtnsStatus` in the places where there is no need to call callbacks. `OtnsStatus` should be called whenever the target status changes, in which case, we should also call callbacks. If the target status does not change, we don't need to call `OtnsStatus` or callbacks. Replacing `OtnsStatus` with callbacks should have the exact same overhead of maintaining code dependency as core code changes. 

---
With all these concerns, I am not sure using registered callbacks is suitable for status emitting to `OTNS`. For now, it seems best and logical to me to integrate `OTNS` utilities to OT core and expect future improvements when we gain more experience.

Reply:
Thanks @simonlingoogle. My goal is understand the approach and help.

> One of the motivation for OTNS to use status emitting to acquire node status rather than calling APIs is because the node status affects the simulation result immediately.
> For example, OTNS uses node rloc16 and extaddr for destination address filtering so that it only dispatches unicast messages to the right destinations. This is one of the reasons that OTNS outperforms existing virtual time simulator by tens of times.

I think this is simulation of radio platform layer, and I agree filtering frames during simulation and dispatching the frame only to the actual dest  is a very good improvement and optimization. 

One thing that is not clear to me is why we would need the OT core to emit events about rloc16 and/or extended address changes to get such optimization. 
- We already have radio platform APIs that inform radio platform when ext address or short address changed (`otPlatRadioSetExtendedAddress()` and  `otPlatRadioSetShortAddress()`). 
- These APIs are indeed used by radio drivers to change and filter on actual hardware (btw, we have other parameters like channel and PAN Id which are also set by OT on radio platform layer and used for frame filtering during rx).
-  OTNS is expected to simulate the radio (and rest of platform layer). If OTNS needs events such as  "short/ext address changed" to update how it would filter/behave, I would say we should emit such events from the platform layer implementation of these APIs and not from OT core. 
- This would be actually immediate and no issue with ordering (or waiting for posted tasklets).
- When OT core calls `otPlatRadioSetShortAddress()` is the exact moment from which the radio simulation should start using the new given short address and node receive only frames for that address. So the ONTS event should be emitted from this function call vs. emitting it from MLE layer method `Mle::SetRlco16()` which is technically  many layers higher in the network stack.

Is there any reason why we cannot do this from the platform layer code and need it to be emitted from OT core?

-----

I would say for any state/parameter changes that is needed for simulation of platform layer (e.g., short/ext address, pan id, channel, etc), we should already have a platform API to set it on platform layer. We should be able to emit events to OTNS simulator from the platform implementation.

I guess/think other parameter/state changes (say, role or partition id, child attached etc) which are not directly used/needed for simulation of platform layer, are intended for visualization by OTNS. Is that correct? 
- If so, does the ordering of events matter for visualization? 
- I feel it should not and then I would suggest using existing notification model (`Notifier` and state changed callbacks) and/or public APIs to query for visualization purpose 
  -  Of course we should improve/add any missing features. 
  - For example if certain changes that would be useful for visualization don't have a related `OT_CHANGED` events we can add them;
  - Or enhance the Child/Neighbor table callbacks to allow more than one registred callback so that one can be used by OTNS for visualization).

Reply:
> Is there any reason why we cannot do this from the platform layer code and need it to be emitted from OT core?

Yes, we can implement the status emitting to `OTNS` in platform code, after following enhancements to OT core:
* Add missing callbacks that is used by `OTNS`
* Enhance each callback to support multiple callbacks
* Enhance Child/Neighbor callbacks to be able to pass the corresponding `ExtAddress`

However, with such effort and increased code complexity, what we get is a strict binding of `OTNS` to the simulation platform, which makes it difficult to port `OTNS` to other platforms. 

Meanwhile, the current implementation has all logic code ready in OT core and requires only one platform function `otPlatOtnsStatus` to be implemented. For example, we can implement `otPlatOtnsStatus` for `NRF52840 Dongle` to emit the status information through UART to visualize devices that connected to the host computer. 

Overall, I think using callbacks in platform code does not have strong benefits, but impose restriction to future `OTNS` enhancements. 

---
Currently, `OtnsStatus` calls are scattered in OT core, which I think should be refactored.  **It would be better if we can use one core component just for `OTNS` related code.** However there is currently no mechanism for signalling synchronous events among core components.

Reply:
@simonlingoogle , I know it may feel that the current model of emitting `otnsStatus()` from different modules within OT is simple and works fine (and tested), and seems like changing the approach to use `Notifier` and other callbacks is adding complexity and more work, but my advice is it is always better to spend more time at start of a new solution and ensure to have a good approach/design (which is better suited for future) than to try to adapt/change it later on. 

> However, with such effort and increased code complexity, what we get is a strict binding of OTNS to the simulation platform, which **makes it difficult to port OTNS to other platforms**.

I don't follow this point about "hard to port on other platforms". From what I understand OTNS has two aspects, simulation (of platform layer) and visualization (or nodes, etc). I feel we should treat these differently.
- I expect OTNS simulation aspect to  only apply when we want to emulate/simulate the platform layer and in that case we would need to know when short/ext address changes, etc.  So it should be fine to emit the events from `otPlatRadio` implementation under simulation platform. 
- Visualization aspect I guess can be used with simulated nodes or with nodes running on actual hardware but if we are using actual hardware then their radio driver would provide MAC filtering and we would not need to know when short/ext address changes. The rest of events needed for visualization only can still use `otnsStatus` model and be common for all platforms.
  - If in future you are considering hybrid cases where some nodes are simulated and some are actual device, we need to update the radio platform code for this anyways and can still emit events for OTNS from the `otPlatRadio` functions. 

> Enhance Child/Neighbor callbacks to be able to pass the corresponding ExtAddress

The [`otNeighborTableCallback` mechanism](https://github.com/openthread/openthread/blob/35bb0027d0b6d2b1b2e54a289bb3e3f56def57d0/include/openthread/thread_ftd.h#L599-L631) already provides full info about neighbor/child being added or removed as `otChildInfo`/`otNeighborInfo` entry (which contains ext address and many other parameters) and it is an immediate callback. 


----------------

Here is my suggestions on how to hook OTNS into OT:

We should divide the events into the ones needed for simulation and the ones for visualization and use different approach for them.

For events required for simulation, I suggest two options:
1. We can use `otPlat...` calls and emit the event to OTNS from them  (this would be my preferred approach).
1. If you still want it to be in core, I would suggest to limit calls to OTNS for simulation from lowest layer of code in OT, for example from the `Radio.hpp` method (e.g., `Radio::SetShortAddress()` for emitting short address change event to OTNS).

For events for visualization (role change, joiner state change, child/neighbor added, etc)
- We can make OTNS a utility like module in under `src/core/util`. 
   - It can either register a callback to `Notifier` and from callback it can emit the events in the string format required by OTNS.
   - If we do really need immediate callback (cannot wait for posted tasklet of Notifier mechanism - which I expect we should not need for visualization), the approach I would suggest is to add a hook in `Notifier::Signal()` to call a method in `OTNS` which would then immediately process the event and emit the related string format event to OTNS. This way we limit the interaction of OTNS with other OT_CHANGED ones (which are tied to public APIs of OT).
-  For neighbor/child table updates, I see two options:
   - Update the [`otNeighborTableCallback` mechanism](https://github.com/openthread/openthread/blob/35bb0027d0b6d2b1b2e54a289bb3e3f56def57d0/include/openthread/thread_ftd.h#L599-L631) to allow more than one callback and register a callbacks from OTNS and process/emit events from there .
   - Or adding a hook in `MleRouter::Siganl()` method to call  into OTNS (note that `Signal()` already has a reference to `Neighbor` entry being added/removed).

Reply:
@abtink any further comments on the new implementation?

Reply:
The new model looks great. Thanks 👍